module #2. REST API Basics/rest-api-basic/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,241 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.dto.GiftCertificateDTO;
+import com.epam.esm.dto.TagDTO;
+import com.epam.esm.exception.model.GiftCertificateAlreadyExistsException;
+import com.epam.esm.exception.model.GiftCertificateNotFoundException;
+import com.epam.esm.model.GiftCertificate;
+import com.epam.esm.model.Tag;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.repository.util.QueryParams;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.service.mapping.MappingService;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.Validate;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.TransactionStatus;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.transaction.support.TransactionCallbackWithoutResult;
+import org.springframework.transaction.support.TransactionTemplate;
+
+import java.time.LocalDateTime;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * This class implements functionality of operating {@link GiftCertificateRepository}
+ * and {@link TagRepository} methods in according to received
+ * parameters from GiftCertificate controller
+ */
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+    private final GiftCertificateRepository giftCertificateRepository;
+    private final TagRepository tagRepository;
+    private final MappingService<GiftCertificate, GiftCertificateDTO> certificateMappingService;
+    private final MappingService<Tag, TagDTO> tagMappingService;
+
+    private final TransactionTemplate transactionTemplate;
+
+    @Transactional
+    @Override
+    public void save(GiftCertificateDTO giftCertificateDTO) {
+        if (giftCertificateRepository.isExists(certificateMappingService.mapFromDto(giftCertificateDTO))) {
+            log.error("[GiftCertificateService.save()] GiftCertificate with given name:[{}] already exists.",
+                    giftCertificateDTO.getName());
+            throw new GiftCertificateAlreadyExistsException(String.format(
+                    "GiftCertificate with given name:[%s] already exists.", giftCertificateDTO.getName()));
+        }
+        GiftCertificate giftCertificate = certificateMappingService.mapFromDto(giftCertificateDTO);
```

Comment:
`certificateMappingService.mapFromDto(giftCertificateDTO)` is used the second time in this method. Please extract it and assign the result to an variable